### PR TITLE
ci: Enable tests to run on forks and action-created PR

### LINKED ISSUE #845: CI: tests do not run on forks or "Version Packages" PR
`Tests` workflow do not run on forks or "Version Packages" PR because...

1. Workflows triggered from forks do not have access to environment secrets for security reasons.
2. Workflows triggered from PRs [do not run](https://github.com/peter-evans/create-pull-request/issues/48) when a PR is created by Github actions.

Reading through https://securitylab.github.com/research/github-actions-preventing-pwn-requests/, we can leverage label-triggered e2e test that acts as "approve and run" button.

### PR DESCRIPTION
**Pull request by wlee221**
*Issue #, if available:* Closes #845

### Description of changes 

This enables **maintainers** to run tests on forks and PRs created by `github-actions` like #838. Tests on forks couldn't run because environment secrets aren't passed for security reasons.

To enable tests on these cases, `tests.yml` trigger is updated to `pull_request_target`. It is granted a read/write repository token and can access secrets even when it is triggered from a fork, so we need to be very careful about the tokens/secrets we expose.

https://securitylab.github.com/research/github-actions-preventing-pwn-requests/ has practical advices. This PR uses label-based approach -- maintainer first checks that code changes do not tamper secrets/tokens, then they can add `run-tests` to trigger the tests. Tests will run only when the label was *added* and not on subsequent commits to prevent malicious commits after the label was added.

### Additional security hardening:

- [x] `GITHUB_TOKEN` should only have write access to `pull-requests`
- [x] Workflow will automatically remove `run-tests` label after the workflow was triggered.
- [x] #856 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -7,11 +7,39 @@ on:
...
 jobs:
+  setup:
+    runs-on: ubuntu-latest
+    # We run tests only if it's:
+    #   1) push to main, or
+    #   2) pull request not from a fork (ie. internal PR), or
+    #   3) pull request from a fork (ie. external PR) that was added "run-tests" label
+    if: |
+      github.event_name == 'push' ||
+      (github.event.pull_request.head.repo.full_name == github.repository) ||
```
Comment (wlee221) [inline .github/workflows/tests.yml:27]: e.g. https://github.com/wlee221/changeset-poc/pull/29

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -7,11 +7,39 @@ on:
...
+  setup:
+    runs-on: ubuntu-latest
+    # We run tests only if it's:
+    #   1) push to main, or
+    #   2) pull request not from a fork (ie. internal PR), or
+    #   3) pull request from a fork (ie. external PR) that was added "run-tests" label
+    if: |
+      github.event_name == 'push' ||
+      (github.event.pull_request.head.repo.full_name == github.repository) || 
+      (github.event.action == 'labeled' && github.event.label.name == 'run-tests')
```
Comment (wlee221) [inline .github/workflows/tests.yml:28]: e.g. https://github.com/wlee221/changeset-poc/pull/30

File: .github/workflows/tests.yml [hunk 3]
```diff
@@ -126,6 +162,10 @@ jobs:
     steps:
       - name: Checkout Amplify UI
         uses: actions/checkout@v2
+        with:
+          ref: ${{ github.event_name == 'push' && github.ref || github.event.pull_request.head.ref }}
```
Comment (ericclemmons) [inline .github/workflows/tests.yml:166]: Can you add a comment on this logic?

It reads like `github.ref` is correct when we push, but PRs rely on the `head.ref` on what to test?

**Review by ericclemmons — APPROVED** [review]
This looks great!  Let's have one last internal conversation just so there are no surprises. 🙏

File: .github/workflows/tests.yml (hunk 3 above)
Comment (wlee221) [reply .github/workflows/tests.yml:166]: Good point! Yes, that's it. `github.ref` on `pull_request_target` always points to the target branch (`main`), so we need to make it a bit more explicit.

Comment (wlee221) [reply .github/workflows/tests.yml:169]: https://github.com/aws-amplify/amplify-ui/pull/855/commits/76552aa0891bcba68de38fea6cf63af158866b23

**Pull request merged**